app/controlplane/internal/authz/authz.go

```diff
@@ -0,0 +1,158 @@
+//
+// Copyright 2024 The Chainloop Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Authorization package
+package authz
+
+import (
+	"errors"
+	"fmt"
+
+	_ "embed"
+
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"github.com/casbin/casbin/v2/persist"
+	fileadapter "github.com/casbin/casbin/v2/persist/file-adapter"
+	entadapter "github.com/casbin/ent-adapter"
+	"github.com/chainloop-dev/chainloop/app/controlplane/internal/conf"
+)
+
+const (
+	// Actions
+	ActionRead   = "read"
+	ActionList   = "list"
+	ActionUpdate = "update"
+	ActionDelete = "delete"
+
+	// Resources
+	ResourceWorkflowContract = "workflow_contract"
+	ResourceCASArtifact      = "cas_artifact"
+	ResourceReferrer         = "referrer"
+)
+
+// resource, action tuple
+type Policy struct {
+	Resource string
+	Action   string
+}
+
+var (
+	PolicyWorkflowContractList   = &Policy{ResourceWorkflowContract, ActionList}
+	PolicyWorkflowContractRead   = &Policy{ResourceWorkflowContract, ActionRead}
+	PolicyWorkflowContractUpdate = &Policy{ResourceWorkflowContract, ActionUpdate}
+	PolicyArtifactDownload       = &Policy{ResourceCASArtifact, ActionRead}
+	PolicyReferrerRead           = &Policy{ResourceReferrer, ActionRead}
+)
+
+type SubjectAPIToken struct {
+	ID string
+}
+
+func (t *SubjectAPIToken) String() string {
+	return fmt.Sprintf("api-token:%s", t.ID)
+}
+
+//go:embed model.conf
+var modelFile []byte
+
+type Enforcer struct {
+	*casbin.Enforcer
+}
+
+func (e *Enforcer) AddPolicies(sub *SubjectAPIToken, policies ...*Policy) error {
```

Comment:
I am guessing there is no chance for `e` to be nil, right?

Reply:
Technically, there is a chance for a nil receiver. I don't handle it because I think it's pretty idiomatic in this specific scenario to let it panic. For example, look at [this method](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/http/client.go;l=589) in the standard library.

So I'd rather keep it like this, but I'd like to know if you think otherwise.